### DISCUSSION TITLE
Compute shaders in Ebitengine

### DISCUSSION THREAD

**Discussion by audiBookning** [hajimehoshi/ebiten#3080] [Q&A]
Thanks for the engine. It looks great.

My question is, as in the title, Is there a way to use compute shaders in Ebitengine?

In Kage, i see that we can use fragment shaders but compute and other shader types are not an option.

I tried several ways to get the context of the window to access opengl functionality (other gpu API are acceptable) but i did not manage to do it. I really do not want to hack the original engine as i think it would only complicate everything (hacking) with little to gain for it. I even think that i saw some comment at a time about ebitengine using a old api for compatibility reasons that could probably not use compute shaders? Or am i wrong?

So am i correct in thinking that it is not possible to do it now? To get the window context and do custom opengl/vulkan/etc operations as advanced as compute shaders? If so i would ask to consider this feature for the engine.
And, also great, would be to have a very basic example of using such context in the example directory of the repo. I do not care much for it being opengl or anything else, as  long as it is a standard api that can be learned and used with confidence.

In my own use case, i was needing to use compute shaders, not for graphics, but exactly for what they are named: efficient parallel computing.

I understand that accessing such a context would possibly bring some weird bugs if one is not very careful. We would be getting out of the engine guaranties after all. That is what customisation entails.
But i do think that the people that would be using this type of context would be quickly filtered by the problems that it would bring. 
And it would be better if it would be "not recommended" clearly and openly to filter even more people from using it without being aware of the "monsters ahead".

But even after all of this said, i still think that such a feature would give much more power to the engine to do some powerful custom games, that the engine could not do as it is now (or as i think it is now).

Thanks.

**Comment by hajimehoshi** [comment]
Thank you for the suggestion. As Ebitengine is just a 2D game engine, I don't plan to have compute shaders. I don't close this discussion thread so feel free to put your ideas here.

**Comment by audiBookning** [comment]
Thanks for the reply. 

I would just add a little more arguments and say that compute shaders are not about being 3d or anything else. They are about being able to do some efficient parallel computing. For a very basic example, we could do some very fast linear algebra in it that our game could need.
Or for some fluid, particles, cellular automata, etc simulations that could be used in the game.

A famous example is Noita. It is a 2d game but it needs cellular automata capabilities that are pretty unreasonable without using compute shaders. That is why the author has to use a custom engine.

There are many such example. It all depends on the need of the dev.
In the end, compute shaders are just a way to use the parallel power of the gpu beyond the normal use of graphics.

Thanks for taking the time to reply and consider the question.

**Reply by nexovec** [reply]
As far as I'm aware, Noita does its thing on the CPU. Your CPU is a beast, you know.

**Reply by nexovec** [reply]
I do know some of the early mobile games used fragment shaders to do computation, effectively using textures as data buffers. It has several drawbacks, but it can be done. Feel like giving it a try?

**Reply by audiBookning** [reply]
Yes. I  am corrected. Noita was not using compute shaders as far as i am aware. And even if so, its example would not be a good for my argument in ebiten context.
At most it would be a very extreme case where the complexity of doing it gpu based would be a bad investment for most people. So although it is possible and similar ideas have already been done using compute shaders, it was not in the spirit of my proposition.
In my case, i was thinking of some of my own existing projects or prototypes where this would be advantageous.

But there are tons of useful "not dev-time complex" stuff than can be achieved using compute shaders. The advantage is that, no matter how much cpu you have, for certain situations we can only be grateful to access the many unused and dormant parallel cpus that could make our cpu have less burden on its shoulders.

As for "fragment shaders to do computation", the tech is still there. and if it needs to be used then one can revive it. I am sure that there are many that do it for their own reasons.
There are other "convoluted" techniques that are still commonly used like, for example, gpu-based flipbook animation, where the animation is baked in a texture. Which by the way, i think we can do using ebiten? I did not try it.

But to tell the truth, personally at this time i cannot see a immediate practical use case for using shaders for computation with modern hardware capacities. It is so old school that it feels similar to you asking me to use 8 bits assembly to build a 2025 commercial game. 😨 
I probably do not feel like giving a try at this time.

**Reply by nexovec** [reply]
I randomly stumbled upon this to check if this kind of thing is possible, but don't have a use-case either. Maybe it's a good option to try OpenCL or cuda to do compute shaders. And at that point, why bother having that in the engine?

**Reply by audiBookning** [reply]
I do not really have a good answer to that question.
At most i can only point to my first post to give some arguments for it or, 
in extreme, just question directly the "why bother" question in itself?
 
Why bother?
Why bother with anything?
My proposition is just that. A proposition. Just like any other.
There is no do or die story linked to it.
If the engine devs think it is valuable and have the time and conditions to do it, then great.
I think the engine will be potentially better for it.
If not, then all is good.
I think the engine is already great as it is and is getting better by the day.
And there are always alternatives if someone like me really need to do some "out of bounds stuffs".
When i first posted, my intention was more about getting feedback from the dev to this question.
If there was strong interest, i could consider pr for it, when, later on, i began that prospective project that was lobbying for something like compute shaders.
If not then, as i said earlier, all is good.
 
Personally I do not mind supporting and arguing for the proposition. 
That is, until what i see as a reasonable limit.
At this precise time, to me it is certainly not a priority. In Ebitengine or in any other form.
Time is hard to stretch for anything these days.
And i think it is possibly the same for the engine dev.

**Reply by nexovec** [reply]
I'd usually bother if someone relied on the feature.
If you start programming just because you can without knowing what you want, what you get is cpython.

I'd be really interested in getting vertex shaders through kage first. I can't even really figure out what sort of 2D game you'd have to be making in order to need compute shaders. Maybe with some giant fluid and gas simulation sandbox.

**Comment by nexovec** [comment]
> I do know some of the early mobile games used fragment shaders to do computation, effectively using textures as data buffers. It has several drawbacks, but it can be done. Feel like giving it a try?

It just so happens that a video called "Graphics vs Compute Pipeline" by Cherno just aired, which showcases how you can emulate a compute shader via a fragment shader in openGL. It should also be possible to do the same thing in ebiten right now.
